### ISSUE TITLE
grant_type=refresh_token doesn't replace existing refreshToken

### ISSUE THREAD

**Issue by gonzalad** [spring-projects/spring-security-oauth#816]
I'm using Spring OAuth 2.0.10.

I have an OAuth Application Client.
When my access token (AT) expires, Spring OAuth uses the refresh token (RT) to get a new AT from the Authorization Server (AS).
Problem is if the AS returns a new RT along the new AT, Spring OAuth doesn't pick the new RT and continues using the old RT.
So on next refresh, I'll get an error.

This is due to https://github.com/spring-projects/spring-security-oauth/blob/b055772adbb9c10497b1077a0480bf2c2649fb04/spring-security-oauth2/src/main/java/org/springframework/security/oauth2/client/token/AccessTokenProviderChain.java#L166.

It should check if there's a RT in the AS response.
If no, just keep using the old RT. If yes, use the new RT.

In my scenario, here's the request to the token endpoint :  

```
POST /oidc/oauth2/token HTTP/1.1
grant_type=refresh_token&refresh_token=90a6e10e2801925583b69ba076c195
```
And the response : 
```
{
   "access_token":"b6f54d7326354cdfb12227c96b6894a",
   "token_type":"Bearer",
   "expires_in":3600,
   "scope":"openid hello.say",
   "refresh_token":"e2f9d5d05ed793d2b096a59dcde10e2",
   "id_token":"eyJhbGciOiJSUzI1NiJ9.eyJwcmVmZXJyZWRfdXNlcm5hbWUiOiJhbGljZSIsInN1YiI6IkNHUjBXTEpaNFdEb1p6ZXExM3h3UXciLCJpYXQiOjE0NzAyOTUxNDgsImV4cCI6MTQ3MDI5ODc0NywiaXNzIjoiYWNjb3VudHMudGFsZW5kLmNvbSIsImdpdmVuX25hbWUiOiJhbGljZSIsImZhbWlseV9uYW1lIjoiYWxpY2UiLCJlbWFpbCI6ImFsaWNlQGdtYWlsLmNvbSIsIm5hbWUiOiJhbGljZSBhbGljZSIsImF1ZCI6IkRCUW91c1N5OFpLNEFBIiwiYXpwIjoiREJRb3VzU3k4Wks0QUEiLCJhdF9oYXNoIjoiU2pyZkF4N2ozM2VvRWF4Vi1JeXRIQSJ9.XxTitaQ5RO7MKfphbKjb0cErBHLDTimj1UTYkYRN1M_EhDnS6TvSHGoryCwMhQ_yANdWOeqZeX_IUmQu2Wj549d_HpLAqzlzrzc1SAbd0nucjWoH3tcCZ2WsuOLigZvhJZpvyj-v-Z2Ov8usk_-fXAP0e-p-MpvGinQ3op-BJQARq2ktM07hQTs8tcVaekpMrjplu-MlxNmqeKSMOQI2TERFZnQRfIYFTQvbCm7hxIs1YJoHnFAh2R_TA_14AJYMBRKXI1yi_EvdJvBkfh-3b7Dxr4pIRIty6aRcsys_9jqF0y4Wdw9Dq6R3GgWgSgvOuyDxq5rmVaR7jLkSw2HVDw"
}
```

Spring OAuth keeps using the old RT (90a6e10e2801925583b69ba076c195) and not the new one (e2f9d5d05ed793d2b096a59dcde10e2).

Extract from OAuth2 specs : 
See https://tools.ietf.org/html/rfc6749#section-6
> The authorization server MAY issue a new refresh token, in which case the client MUST discard the old refresh token and replace it with the new refresh token.  The authorization server MAY revoke the old refresh token after issuing a new refresh token to the client.  If a new refresh token is issued, the refresh token scope MUST be identical to that of the refresh token included by the client in the request.

**Comment by dsyer** [comment]
Coincidentally, I noticed this yesterday as well. This behaviour came in the fix for #715 I believe.

**Comment by agokulravikumar** [comment]
This issue seems to be exists and not resolved. 

I am using the latest **2.4.0 RELEASE** and using **JDBCTokenStore**

When I have the **grant_type** as **refresh_token**, for the first request everything works fine. But from the next refresh token request it throws invalid refresh token error.

In my analysis found that, during refresh_token request, the token_id in oauth_refresh_token table is not getting replaced. So it creates a mismatch between **oauth_access_token.refresh_token** and **oauth_refresh_token.token_id**

Also, when I try refresh_token request with the first created refresh token it is still responding with new access token each time which is a **serious security threat**

Due to this multiple access token are generated in oauth_access_token table for the same username.

Can someone update me on this ?
